xonsh/prompt/vc_branch.py

```diff
@@ -64,20 +65,27 @@ def _get_parent_dir_for(path, dir_name, timeout):
             return path
         previous_path = path
         path, _ = os.path.split(path)
-    return (path == previous_path)
+    return path == previous_path
 
 
-def get_hg_branch(cwd=None, root=None):
+def get_hg_branch(root=None):
+    """Try to get the mercurial branch of the current directory,
+    return None if not in a repo or subprocess.TimeoutExpired if timed out.
+    """
     env = builtins.__xonsh_env__
-    cwd = env['PWD']
-    root = _get_parent_dir_for(cwd, '.hg', env['VC_BRANCH_TIMEOUT'])
-    if not isinstance(root, str):
-        # Bail if we are not in a repo or we timed out
-        if root:
+    timeout = env['VC_BRANCH_TIMEOUT']
+    # Bail if we are not in a repo or we timed out
+    try:
+        status = subprocess.call(['hg', 'status'], timeout=timeout,
+                                 stdout=subprocess.DEVNULL,
+                                 stderr=subprocess.DEVNULL)
+    except subprocess.TimeoutExpired:
+        return subprocess.TimeoutExpired(['hg'], timeout)
+    else:
+        if status == 255:
             return None
-        else:
-            return subprocess.TimeoutExpired(['hg'], env['VC_BRANCH_TIMEOUT'])
-    if env.get('VC_HG_SHOW_BRANCH') is True:
+    root = _get_parent_dir_for(env['PWD'], '.hg', timeout)
```

Comment:
This is still doing the same call to get the root. The root is used to get the current branch and bookmark from the ```.hg/``` directory.

This has just introduced an ```hg status``` call.

I assume that you were getting branch timeouts due to the current directory being many sub directories from the hg root.

But I don't see how this would help since this is still doing the same ```_get_parent_dir_for()``` call.

Ideally the code would use the ```hg root``` command. I found it faster to not invoke ```hg``` but would be happier with that instead of doing ```hg status``` and then again finding the root directory.

I personally use a different Python to run ```hg``` than I use to run ```xonsh```.